docs/modules/record-encryption/proc-configuring-record-encryption-filter.adoc

```diff
@@ -35,14 +35,21 @@ filters:
       selector: <KEK_selector_service_name> # <3>
       selectorConfig:
         template: "KEK_${topicName}" # <4>
+      experimental:
+        "encryptionDekRefreshAfterWriteSeconds": 3600 # <6>
+        "encryptionDekExpireAfterWriteSeconds": 7200 # <7>
+        "maxEncryptionsPerDek": 5000000 # <8>
 	 # ...
 ----
 <1> The KMS service name.
 <2> Configuration specific to the KMS provider.
 <3> The Key Encryption Key (KEK) selector to use. The `${topicName}` is a literal understood by the proxy. 
 For example, if using the `TemplateKekSelector` with the template `KEK_$\{topicName}`, create a key for every topic that
 is to be encrypted with the key name matching the topic name, prefixed by the string `KEK_`.
-<4> The template for deriving the KEK, based on a specific topic name. 
+<4> The template for deriving the KEK, based on a specific topic name.
+<5> How long to use a DEK for encryption after it has been created, after this time the Filter will asynchronously obtain a new DEK while continuing using the old one. This can be used to control how quickly we respond to the fact that a KEK has been rotated externally. Note that tuning this value has a cost implication. Shorter durations will result in increased operations on the KMS at encrypt and decrypt time, because more DEKs will be created and used.
+<6> How long after creation a DEK is eligible to be evicted from the cache. Only actively used DEKs will be refreshed, this setting puts a bound on how long unused DEKs remain cached.
```

Comment:
> How long after creation a DEK is eligible to be evicted from the cache.

I think the "eligible to be evicted" part is not helpful.  From the user's point of view, the DEK is removed.  The fact that Caffeine holds on to the entries until the maintenance isn't really relevant to the user.

>  Only actively used DEKs will be refreshed, this setting puts a bound on how long unused DEKs remain cached.

By "use" do you mean "access"?  I don't think use/access is relevant here.  We are only interested in "write" times.

```suggestion
<6> How long after creation a DEK until it is removed from the cache. This setting puts an upper bound on how long a DEK can remain cached.
```

Reply:
I had a misconception that since the maintenance is periodic:

> Expiration is performed with periodic maintenance during writes and occasionally during reads. Scheduling and firing of an expiration event is executed in amortized O(1) time.

that it would not be guaranteed to be expired exactly at 2h00m and the cache could still return it until eviction.

But  I have confirmed there is code for async and non-async cache implementations that ensures the old value is not returned to clients and your timeline is accurate.